### ISSUE TITLE
kubelet not call unstagevolume when it restarts after unpublishvolume

### ISSUE THREAD

**Issue by smileusd** [kubernetes/kubernetes#85476]
/kind bug
/sig storage
**What happened**:
When user deletes pod and csi pvc, kubelet will send unpublishvolume and unstagevolume request to csidriver. But when finish unpublishvolume response finished and pod was deleted, kubelet down and restart. Kubelet will lost the information of mount with this volume and it will never to call unstagevolume, and csi-provisioner will delete volume failed forever.
**What you expected to happen**:
Kubelet restart and csi pv deleted.
**How to reproduce it (as minimally and precisely as possible)**:
Delete pvc, when csidriver finish unpublishvolume, restart kubelet.
**Anything else we need to know?**:
```None```
**Environment**:
- Kubernetes version (use `kubectl version`):
1.16

**Comment by smileusd** [comment]
@cofyc @msau42 anyone take a look?

**Comment by jsafrane** [comment]
This is a long known bug - when kubelet starts, it scans only pod directories and reconstructs volume there.

If a volume was unmounted (`TearDown` in volume plugins, `NodeUnpublish` in CSI) and kubelet died before `UnmountDevice` / `NodeUnstage`,  kubelet looses any track of the volume and will not continue with the cleanup after it's restarted.

@msau42, do I remember correctly we decided to ignore this issue for now? Otherwise kubelet needs a real database (or checkpoints), IMO it's not possible to reconstruct volumes from global dirs.

**Comment by hoyho** [comment]
> This is a long known bug - when kubelet starts, it scans only pod directories and reconstructs volume there.
> 
> If a volume was unmounted (`TearDown` in volume plugins, `NodeUnpublish` in CSI) and kubelet died before `UnmountDevice` / `NodeUnstage`, kubelet looses any track of the volume and will not continue with the cleanup after it's restarted.
> 
> @msau42, do I remember correctly we decided to ignore this issue for now? Otherwise kubelet needs a real database (or checkpoints), IMO it's not possible to reconstruct volumes from global dirs.

Can that persist to configmap or etcd?

**Comment by smileusd** [comment]
Thanks @jsafrane. We have persistent data in pods direcectory, why we could not also persistent data in  plugin directory? Or persistent in configmap of csi owner.

**Comment by jsafrane** [comment]
> Can that persist to configmap or etcd?

No, API server may not be reachable at the point when kubelet starts.

> We have persistent data in pods direcectory, why we could not also persistent data in plugin directory? 

PRs are welcome. It's not so easy to implement global directory reconstruction. Even reconstruction from pod directory took us several iterations to get it right. Maybe using some sort of checkpoints could help.

**Comment by smileusd** [comment]
@jsafrane Thanks a lot. Your comments very helpful.

**Comment by msau42** [comment]
/reopen
Maybe we could do something for csi since the staging directory is deterministic.

@kubernetes/sig-storage-bugs
